src/Client/ClickHouseClient.cpp

```diff
@@ -0,0 +1,286 @@
+#include <Common/NetException.h>
+#include <Client/ClickHouseClient.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int DEADLOCK_AVOIDED;
+extern const int TIMEOUT_EXCEEDED;
+extern const int UNKNOWN_PACKET_FROM_SERVER;
+extern const int UNEXPECTED_PACKET_FROM_SERVER;
+}
+
+namespace
+{
+
+std::unique_ptr<Connection> createConnection(const ConnectionParameters & parameters)
+{
+    auto ret = std::make_unique<Connection>(
+        parameters.host,
+        parameters.port,
+        parameters.default_database,
+        parameters.user,
+        parameters.password,
+        parameters.quota_key,
+        "", /* cluster */
+        "", /* cluster_secret */
+        "TimeplusProton",
+        parameters.compression,
+        parameters.security);
+
+    ret->setCompatibleWithClickHouse();
+    return ret;
+}
+
+size_t calculatePollInterval(const ConnectionTimeouts & timeouts)
+{
+    const auto & receive_timeout = timeouts.receive_timeout;
+    constexpr size_t default_poll_interval = 1'000'000; /// in microseconds
+    constexpr size_t min_poll_interval = 5'000; /// in microseconds
+    return std::max(min_poll_interval, std::min<size_t>(receive_timeout.totalMicroseconds(), default_poll_interval));
+}
+
+}
+
+ClickHouseClient::ClickHouseClient(ConnectionParameters params_, Poco::Logger * logger_)
+    : params(params_)
+    , connection(createConnection(params))
+    , poll_interval(calculatePollInterval(params.timeouts))
+    , logger(logger_)
+{
+}
+
+void ClickHouseClient::reset()
+{
+    cancelled = false;
+    processed_rows = 0;
+    server_exception = nullptr;
+}
+
+void ClickHouseClient::executeQuery(const String & query, const String & query_id, bool fail_quick)
```

Comment:
Using this high level connection to send select query works but probably slow for ingestion. I was thinking  we could use low level api of connection like `Connection::sendData(...)` for ingestion.

A very good source code we can refer is `StorageDistributed.cpp`, check out is read / write function and you can follow the rabbit hole to get DistributedSink, RemoteSink, RemoteInserter, Interpreter / Cluster, Connection Pooling and how everything compile together to make it efficient by using low level APIs.

Reply:
StorageDistributed essentially proxy query to remote ClickHouse but at low level with greater efficiency than a plain text send which is not native.

Reply:
Created a ticket for such improvement: https://github.com/timeplus-io/proton/issues/551